File: src/aufgabe6/Person.java
package aufgabe6;

import java.util.LinkedList;
import java.util.List;

public class Person {
    final String name;
    private final List<Buch> ausgelieheneBuecher = new LinkedList<>();

    public Person(String name){
        this.name = name;
    }

    public void print(){
        System.out.println(name + " hat ausgeliehen: " + ausgelieheneBuecher);
    }

    public int getAnzahlAusgelieheneBuecher(){
        return ausgelieheneBuecher.size();
    }

    public boolean leihtAus(Buch b){
        if((b.getEntleiher() != null && b.getEntleiher() != this) || ausgelieheneBuecher.contains(b)){
            return false;
        }
        ausgelieheneBuecher.add(b);
        b.wirdAusgeliehen(this);
        return true;
    }

    public boolean gibtZurueck(Buch b){
        if(!ausgelieheneBuecher.contains(b)){
            return false;
        }
        ausgelieheneBuecher.remove(b);
        b.wirdZurueckGegeben();
        return true;
    }

    public String getName(){
        return name;
    }

    @Override
    public String toString(){
        return name;
    }
}

File: src/aufgabe06/cpp/person.cpp
#include "buch.hpp"

#include "person.hpp"
#include <algorithm>
#include <iostream>

namespace aufgabe6 {

Person::Person(const std::string &name) : name(name) {}

void Person::print() const {
  std::cout << name << " hat ausgeliehen: ";

  if (ausgB.empty()) {
    std::cout << "[]";
  } else {
    std::cout << "[";
    std::for_each(ausgB.begin(), ausgB.end(),
                  [](const Buch *b) { std::cout << b->getName() << ", "; });
    std::cout << "]";
  }
  std::cout << std::endl;
}

size_t Person::getAnzahlAusgelieheneBuecher() const { return ausgB.size(); }

bool Person::leihtAus(Buch *b) {
  if ((b->getEntleiher() != nullptr && b->getEntleiher() != this) ||
      (std::find(ausgB.begin(), ausgB.end(), b) != ausgB.end()))
    return false;
  ausgB.push_back(b);
  b->wirdAusgeliehen(this);
  return true;
}
bool Person::gibtZurueck(Buch *b) {
  if (!(std::find(ausgB.begin(), ausgB.end(), b) != ausgB.end()))
    return false;
  ausgB.remove(b);
  b->wirdZurueckGegeben();
  return true;
}
std::string Person::getName() const { return name; }

std::ostream &operator<<(std::ostream &os, const Person &p) {
  return os << p.name;
}

} // namespace aufgabe6
